### DISCUSSION TITLE
Updates on repo release workflow

### DISCUSSION THREAD

**Discussion by captainsafia** [nteract/archived-desktop-app#5568] [General]
Hello folks,

After successfully getting semantic-release working on the Lerna mono repo we use for outputs (https://github.com/nteract/outputs), I've been able to configure the same for this monorepo.

From now on, packages will be automatically released if changes have been made to a particular area.

Coupled with this, I've also added back a precommit linter to validate that commits on the repo follow the conventional-commit style so that releases will be versioned appropriately.

There's still some work that needs to be done to support writing back updated version numbers to the repo that I'll do.

One step closer to more automation!

**Comment by captainsafia** [comment]
Alrighty -- we've got the first automated release coming out of the monorepo in the editor package (see https://github.com/nteract/nteract/commit/e75829818415f4b9777ca936754b5e33a66bf876).

What will the bot do? Whenever a PR is merged, it will audit the commits pushed to a particular package area between the commit associated with the last released tag and the current commit. Based on the commit conventions, it'll determine what package version to release and release it. It'll also submit a PR back to update the package.json for the file.

💡  Note: unlike the lerna release process, it will not update the package.json for dependents of the package. This is mostly fine though since we pin to minor versions on all monorepo references (e.g. "@nteract/commutable": "^1.2.3") so patch and minor releases will be handled correctly. For major releases, we probably want to go through and push it out in a separate commit so the right things happen.

⚠️ Alert: this change means we have to be very prudent about our commits. I've added a new pre-commit hook that audits messages to ensure they follow the [conventionalcommit](https://www.conventionalcommits.org/en/v1.0.0/) style but devs will also have to make sure that fixes, features, and breaking changes are labelled correctly.

With this in place, here are some next steps:

- [x] Figure out how to get semantic-release to produce the changelog/release notes for each package
- [x] Trigger one last manual release with all the packages
- [x] Audit any unmerged PRs at this point to ensure they follow the conventions
- [x] Update the contributing and release docs to reflect the new process
- [ ] Work on getting automated releases for the desktop app

**Reply by rgbkrk** [reply]
Thanks for committing us to this process. I'll commit to committing conventional commits.

**Reply by hydrosquall** [reply]
Thanks for this write-up! I used it to set up something similar for data-explorer: https://github.com/nteract/data-explorer/pull/66#issue-681340816

One thing I'm considering adding is a CLI tool to help people generate commits that follow these guidelines, specifically `commitizen`: https://github.com/commitizen/cz-cli#if-your-repo-is-commitizen-friendly . If it works out in data-explorer, it may be useful for nteract core too.

**Reply by captainsafia** [reply]
Yep! I've added a linter as a pre-commit step. It doesn't quite guide commit creation like commitzen does but it ensures that the correct commit format is used before a user creates a commit.

**Reply by hydrosquall** [reply]
Nice! I've added the linter as a pre-commit step (#66) + added commitizen too ( https://github.com/nteract/data-explorer/pull/68 )
